Documentation for the Docker Compose Application Update Module

alfrunes opened the pull request.

alfrunes commented:
Requires mendersoftware/app-update-module#31 (app-gen fix) and mendersoftware/app-update-module#30 (update module fixes) to work.

TheMeaningfulEngineer commented on 01.Get-started/04.Deploy-a-container-update/docs.md:
```diff
@@ -186,8 +186,12 @@ The [Docker Update Module](https://hub.mender.io/t/docker/324?target=_blank)
+There are several options for deploying containerized updates:
+
+* *[Docker Compose Update Module](../../06.Artifact-creation/03.Containerized-application-updates)* provides an interface for deploying container workloads managed by Docker Compose.
+* *[Kubernetes Update Module](https://hub.mender.io/t/kubernetes/1939?target=_blank)* allows you to deploy Kubernetes manifests to a local kubernetes cluster.
```
I would scrap this example completely.
If we now have a supported version on how to do a container update, let's not have a getting started as a community based one.

TheMeaningfulEngineer replied:
Two actions in one comment, reducing to only one.

Community update modules shouldn't listed in the same was as supported ones.

TheMeaningfulEngineer replied:
Let's just remove Kuberenetes

TheMeaningfulEngineer commented on 06.Artifact-creation/03.Containerized-application-updates/docs.md:
```diff
@@ -0,0 +1,238 @@
+ * [xdelta3](https://github.com/jmacd/xdelta)
+   * Only required if using binary deltas
+
+[ui-accordion independent=true open=none]
```
This looks nice, but not obvious from the rendering it was a button.
Would just keep it expanded.

Vefify -> Verify

TheMeaningfulEngineer replied:
Deleting, Lucho commented as well.

TheMeaningfulEngineer commented on 06.Artifact-creation/03.Containerized-application-updates/docs.md:
```diff
@@ -0,0 +1,238 @@
+ * [xdelta3](https://github.com/jmacd/xdelta)
+   * Only required if using binary deltas
+
+[ui-accordion independent=true open=none]
```
This looks nice, but not obvious from the rendering it was a button.
Would just keep it expanded.

Vefify -> Verify

TheMeaningfulEngineer commented on 06.Artifact-creation/03.Containerized-application-updates/docs.md:
```diff
@@ -0,0 +1,238 @@
+---
+
+Mender maintains and support update modules for managing containerized workloads
+on the edge devices. The Application Update Module extends the update module API
```
What's an Application Update Module?
The fact that Docker compose is based on it doesn't tell me much about the Application Update module.

TheMeaningfulEngineer replied:
Needs a chapter with a definition of an Application Update module.

TheMeaningfulEngineer commented on 06.Artifact-creation/03.Containerized-application-updates/docs.md:
````diff
@@ -0,0 +1,238 @@
+        -O /etc/mender/mender-app-docker-compose.conf
+```
+
+!!! Inspect the configuration files on your device `/etc/mender/mender-app.conf` and
````
Based on what do I know how to configure it?

TheMeaningfulEngineer replied:
Section on configuration of the both the Application Update and Container Update Module needed

alfrunes replied:
This was a bit confusing to me too. It should work out of the box, so I'm considering to remove this tip.

TheMeaningfulEngineer commented on 06.Artifact-creation/03.Containerized-application-updates/docs.md:
```diff
@@ -0,0 +1,238 @@
+
+To generate the artifact, we need to know the target platform of the devices we
+want to deploy to. In the following example, we will assume the platform we are
+deploying to is `linux/arm/v7`.
```
Why do we need to know that?
How do I figure out what mine is if I don't know?

MuchoLucho replied:
I would assume a person who follows this know about the platform, my concern is the expected format `x/x/x`. Are we documenting it somewhere else?

TheMeaningfulEngineer replied:
Not actionable, open ended question.

Add part explaining how to get the value for the platform (`linux/arm/v7`).

TheMeaningfulEngineer commented on 06.Artifact-creation/03.Containerized-application-updates/docs.md:
```diff
@@ -0,0 +1,238 @@
+ARTIFACT_NAME="myfirstcomposition"
+DEVICE_TYPE="raspberrypi4"
+PLATFORM="linux/arm/v7"
+app-gen --artifact-name "$ARTIFACT_NAME" \
```
Gave me an error:
```
/home/alan-nt/bin/app-gen: line 78: $1: unbound variable
```

TheMeaningfulEngineer replied:
Ref to closed conversation: https://github.com/mendersoftware/mender-docs/pull/2134/files#r1232426335

TheMeaningfulEngineer replied:
Seems like there are [code dependencies](https://github.com/mendersoftware/mender-docs/pull/2134#issuecomment-1591491346) to be able to verify these docs.
The docs can't be merged until we can successfully execute the snippets.

alfrunes replied:
https://github.com/mendersoftware/app-update-module/pull/32

MuchoLucho commented on 06.Artifact-creation/03.Containerized-application-updates/docs.md:
```diff
@@ -0,0 +1,238 @@
+ * [Mender client](../../03.Client-installation/02.Install-with-Debian-package) (version >= 3.0)
+ * [Docker Engine](https://docs.docker.com/engine/install/?target=_blank)
+ * [Docker Compose](https://docs.docker.com/compose/install/?target=_blank) (version >= 2.0)
+ * [xdelta3](https://github.com/jmacd/xdelta)
```
A little bit curious on how the following commands are going to work if we have a read only rootfs. Should we state you must have a RW rootfs before proceeding if you made it "delta ready" in the past?

TheMeaningfulEngineer replied:
Doesn't sound actionable - pls think in terms of making a PR with improvements you have to suggest instead of open question.

So instead of `Should we state you must have a RW rootfs before proceeding if you made it "delta ready" in the past?` more like `I tested the on a RW file system - didn't work - here is a suggestion for how the docs should address this`.

estape11 commented on 06.Artifact-creation/03.Containerized-application-updates/docs.md:
```diff
@@ -0,0 +1,238 @@
+In this section we will show you how to use the Docker Compose Application
+Update Module for deploying a simple Docker composition on your device.
+
+#### Prepare the device
```
### WIP - CE Docs review

We could be more verbose regarding which device, from instructions I got 2 devices involved, the actual remote device and another one for the artifact generation. This is important as the requirements for each are different, i.e the one in charge for artifact generation is not required to have mender installed.

MuchoLucho commented on 06.Artifact-creation/03.Containerized-application-updates/docs.md:
```diff
@@ -0,0 +1,238 @@
+   * Only required if using binary deltas
+
+[ui-accordion independent=true open=none]
+[ui-accordion-item title="Vefify device dependencies"]
```
Small typo in Verify

MuchoLucho commented on 06.Artifact-creation/03.Containerized-application-updates/docs.md:
```diff
@@ -0,0 +1,238 @@
+[/ui-accordion]
+
+
+To install the Docker Compose update module, run the following commands on your
```
`Update Module` in uppercase for consistency

TheMeaningfulEngineer replied:
Let's do a check for that on the whole document.

TheMeaningfulEngineer reviewed (changes requested):
@alfrunes 
This and all the comments are happening as part of a CE review - all 3 of as are reading it for the first time and than aligning.
Think about it as a WIP util we reach a conclusion on the change or even suggest a PR so don't freak out by the number of comments :)

**Comments on docs structure**

If we are introducing a section on supported update modules the docs should be restructured.
Instead of listing update modules under artifact creation I'd have a whole section on "Update modules"

```
Update modules
  Create a Delta update Artifact
  Server-side generation of Delta Artifacts
  Create a custom Update Module
  Application Update module
    Docker Compose
  Community update modules
```

MuchoLucho commented on 06.Artifact-creation/03.Containerized-application-updates/docs.md:
```diff
@@ -0,0 +1,238 @@
+BINDIR=$HOME/bin
+mkdir -p $BINDIR
+export PATH=$BINDIR:$PATH
+wget https://raw.githubusercontent.com/mendersoftware/app-update-module/master/gen/app-gen \
```
This script contains hashes like: 

docker.io/library/debian:11,docker.io/library/debian:latest@sha256:a94cd7c7d58f483affd5937853ad4d24caa18cd7c2ec9ef65a9e528dfbc5eb07

Should we worry about it?

estape11 commented on 06.Artifact-creation/03.Containerized-application-updates/docs.md:
```diff
@@ -0,0 +1,238 @@
+Before you can deploy Docker compositions to your devices, you need to install
+the Application Update Artifact Generator. Make sure you have
+[`mender-artifact`](../../10.Downloads#mender-artifact) installed on your system and then install the
+Application Update Artifact Generator:
```
### WIP - CE Docs review
Maybe not strictly limited to this doc entry but `app-gen `seems to fail in my environment (tried bash & zsh), apparently the line 78 needs a change:
```bash
...
while [ $# -ge 1 ] && [ -n "$1" ]; do
...
```
After that, I was able to use the script. ([ref](https://unix.stackexchange.com/a/463038))

TheMeaningfulEngineer replied:
This snippet is just downloading the app?
Did you mean: https://github.com/mendersoftware/mender-docs/pull/2134/files#r1232398222

MuchoLucho replied:
Then I would create the PR for https://github.com/mendersoftware/app-update-module/blob/ee12ba4c869f733af86b74469ead20edaa605866/gen/app-gen#L78 instead

TheMeaningfulEngineer replied:
Closing this as Lucho confirmed this is the error I reported and will continue there as it points to the snippet where the error gets triggered.

alfrunes replied:
https://github.com/mendersoftware/app-update-module/pull/32

MuchoLucho commented on 06.Artifact-creation/03.Containerized-application-updates/docs.md:
````diff
@@ -0,0 +1,238 @@
+To this end, we will leverage the Troubleshoot add-on and start a port-forward
+session using the [mender-cli](TODO).
+```bash
+mender-cli port-forward 1bfcf943-4378-4a4f-bc88-0b4c86cdcc74 8080:8080
````
We should not assume the customer knows that the `id` here means the `device id`. To make sure that visit the reference to https://docs.mender.io/add-ons/port-forward#port-forwarding would be a must and state the following line is an example. Or just change it for `mender-cli port-forward device_id 8080:8080`

TheMeaningfulEngineer replied:
Pls translate into real change suggestion.

MuchoLucho replied:
Change for `mender-cli port-forward device_id 8080:8080`  or state the id above must be replaced with your actual id

estape11 reviewed (commented):
CE Docs Review

MuchoLucho reviewed (commented):
Just a first review

TheMeaningfulEngineer commented on 06.Artifact-creation/03.Containerized-application-updates/docs.md:
```diff
@@ -0,0 +1,238 @@
+!!!!! The last argument to `app-gen` ensures that the artifact is only
+!!!!! installed if `v1` is installed.
+
+The `--deep-delta` flag enables the delta feature which creates a binary delta
```
Missing documentation about Application Update Module.
Mentioned that already in anohter commetn.
Resolving.

alfrunes commented:
Any hints on why the internal link checker cannot resolve the header fragments is appreciated :)

TheMeaningfulEngineer commented:
Not sure I'll be able to review this until Friday.
My minimal threshold for giving it a green light will be:
* The example can be executed on the RPi demo image
* It replaces the community container update module in the Getting started - I see no reason the official one shouldn't take priority

alfrunes commented:
> * The example can be executed on the RPi demo image

The fixes I reported in the initial comment are now merged into master. I was using the RPi 4 demo image to test the snippets.

> * It replaces the community container update module in the Getting started - I see no reason the official one shouldn't take priority

I would like to get @tranchitella input here. There have already been several twists and turns in this ticket, and I'm not too keen on escalating this ticket beyond the requirements without making a follow-up ticket for restructuring the docs.

alfrunes commented:
@TheMeaningfulEngineer @estape11 @MuchoLucho Are there any plans for another round of reviews? I'm a bit lost about what to do with this one.

estape11 reviewed (commented):
Sorry I missed hit submit for this review

TheMeaningfulEngineer commented on 06.Artifact-creation/03.Containerized-application-updates/docs.md:
```diff
@@ -19,27 +19,32 @@ supports Docker Compose orchestrator.
  * [Docker Compose](https://docs.docker.com/compose/install/?target=_blank) (version >= 2.0)
  * [xdelta3](https://github.com/jmacd/xdelta)
-   * Only required if using binary deltas
+   * *Only required if using binary deltas*
```
This was in @estape11 PR whihc was merged.

TheMeaningfulEngineer commented on 06.Artifact-creation/03.Containerized-application-updates/docs.md:
```diff
@@ -0,0 +1,242 @@
+    label: tutorial
+---
+
+Mender maintains and support Update Modules for managing containerized workloads
+on the edge devices. *Application Update Module*, being the regular [Mender Update Module](../08.Create-a-custom-Update-Module),
+contains and implements all the logic behind the deployment of any containerized
+application to a device. The *Application Update Module* extends the Update Module API
+to easily support different kinds of container orchestrators. Currently the module
+supports Docker Compose orchestrator.
```
```suggestion
```

We want to present the docker componse use case.
Don't care about the Application Update Module which is just a means to an end.

alfrunes replied:
Removed the first paragraph and the first heading.

TheMeaningfulEngineer commented on 06.Artifact-creation/03.Containerized-application-updates/docs.md:
```diff
@@ -0,0 +1,242 @@
+to easily support different kinds of container orchestrators. Currently the module
+supports Docker Compose orchestrator.
+
+## Docker Compose
```
```suggestion
```

TheMeaningfulEngineer commented on 06.Artifact-creation/03.Containerized-application-updates/docs.md:
````diff
@@ -0,0 +1,242 @@
+wget https://raw.githubusercontent.com/mendersoftware/app-update-module/master/conf/mender-app-docker-compose.conf \
+        -O /etc/mender/mender-app-docker-compose.conf
+```
+
````
```suggestion
!!! [Application Update Module](https://github.com/mendersoftware/app-update-module/blob/master/docs/README-submodule-api.md#applications-updates) contains the generalized logic for application updates. It is a dependency for the Docker Compose Application Update Module, that is why it's being installed.
```

TheMeaningfulEngineer reviewed (changes requested):
Made the suggestions on the Application Update Module.
It looks just like a means to an end "Doing the container update" and not really important.

alfrunes commented on 06.Artifact-creation/03.Containerized-application-updates/docs.md:
```diff
@@ -0,0 +1,238 @@
+
+##### Verify your deployment
+
+Once deployed, the device will start serving a simple server on port 8080. You
```
I'm failing to reproduce these errors that you report. I did, however, encounter an issue with missing `jq`, so I'll add that to the docs. Could you try again? Maybe the `app-update-module` sources was outdated when you last tested it.

estape11 replied:
Indeed Alf, I retested and now works as expected, thank you very much for your work ⚡

alfrunes commented on 06.Artifact-creation/03.Containerized-application-updates/docs.md:
```diff
@@ -6,10 +6,14 @@ taxonomy:
 
 Mender maintains and support update modules for managing containerized workloads
-on the edge devices. The Application Update Module extends the update module API
+on the edge devices. *Application Update Module* (App Update Module), being the regular [Mender Update Module](../08.Create-a-custom-Update-Module),
```
In the first iteration, we want to receive feedback from customers that have shown interest in an Docker Compose update module. I believe once we have more material it makes sense to split this into a top-level chapter. For now, I believe this is too targeted to stay at the the top level.

estape11 replied:
Excellent then, we (CEs) completed the review for this Docs entry, kind regards

TheMeaningfulEngineer commented on 06.Artifact-creation/03.Containerized-application-updates/docs.md:
```diff
@@ -0,0 +1,244 @@
+
+### Prerequesites
+
+To get started using the Docker Compose Update Module, we need to prepare the
+target devices for deployment and as the workstation for creating the
+deployment. We will start with the devices before returning to the workstation.
```
```suggestion
To get started using the Docker Compose Update Module, we need to prepare the
target devices for accepting the deployment and the workstation for creating the
deployment.```

TheMeaningfulEngineer reviewed (approved):
Seems like Esteban got it to work.
I found it hard to install the requirements on the demo images (Raspberry reference, or QEMU) but that's out of scope.

I read the text structure and it make sense.
Just some minor rewording, but lgtm

alfrunes commented:
@TheMeaningfulEngineer @estape11 @MuchoLucho; I squashed all the commits and fixed the internal links. I also moved the section as a sub-chapter of "Create an artifact"  and renamed the chapter to "Tutorial: Docker Compose update", the rest of the content is the same. Could I get a quick review on the updated docs structure?

TheMeaningfulEngineer commented on 06.Artifact-creation/01.Create-an-Artifact/10.Docker-Compose/docs.md:
```diff
@@ -0,0 +1,244 @@
+
+### Prepare the deployment
+
+In this tutorial, we will make a simple multi-container composition with a
+Traefik gateway and a simple backend service `whoami`. First we will create a
+deployment with a new Docker composition, then we will demonstrate how to
+upgrade the images used in the composition using binary delta to save size and
+data transfer bandwidth for the resulting deployment.
```
```suggestion
```

[WIP CE review]

This looks like an excellent replacement for the introduction text.
It gives an overview of what will happen if you follow along, and gently introduces the delta.

Will suggest it as a replacement for the intro text, in which case this can only focus on the "get it working without xdelta"

estape11 replied:
Hey @TheMeaningfulEngineer xdelta is a must (based on our testing)

alfrunes replied:
Rewrote this section and moved your modified version to the intro (with slight rewording).

TheMeaningfulEngineer commented on 06.Artifact-creation/01.Create-an-Artifact/10.Docker-Compose/docs.md:
```diff
@@ -0,0 +1,244 @@
+upgrade the images used in the composition using binary delta to save size and
+data transfer bandwidth for the resulting deployment.
+
+Begin by creating the manifest and saving it in a separate directory:
```
```suggestion
Begin by creating the manifest on your workstation and saving it in a separate directory:
```

TheMeaningfulEngineer commented on 06.Artifact-creation/01.Create-an-Artifact/10.Docker-Compose/docs.md:
```diff
@@ -0,0 +1,244 @@
+ * [Docker Engine](https://docs.docker.com/engine/install/?target=_blank)
+ * [Docker Compose](https://docs.docker.com/compose/install/?target=_blank) (version >= 2.0)
+ * [`jq`](https://jqlang.github.io/jq/)
+ * [`xdelta3`](https://github.com/jmacd/xdelta)
```
WIP CE review
We used to have a mention that this was optional.
A reader can achieve the first part of the update without worrying about xdelta so let's not insist on it up front

estape11 replied:
@TheMeaningfulEngineer, from our testing it was required (the update module fails if is not installed)

TheMeaningfulEngineer reviewed (changes requested):
Looks good.
Curretly reviweing, will sync with other CEs to see if we can just make all these comments into a PR instead of you having to individually adress them.

estape11 commented on 06.Artifact-creation/01.Create-an-Artifact/10.Docker-Compose/docs.md:
```diff
@@ -0,0 +1,244 @@
+    label: tutorial
+---
+
+In this section we will show you how to use the Docker Compose Application
+Update Module for deploying a simple Docker composition on your device.
```
@TheMeaningfulEngineer it is truncated (..composition with a), and you do you intend to suggest a list for this?

estape11 commented on 06.Artifact-creation/01.Create-an-Artifact/10.Docker-Compose/docs.md:
```diff
@@ -0,0 +1,244 @@
+deployment with a new Docker composition, then we will demonstrate how to
+upgrade the images used in the composition using binary delta to save size and
+data transfer bandwidth for the resulting deployment.
+
```
```suggestion
##### Create the Mender artifact
```

estape11 commented on 06.Artifact-creation/01.Create-an-Artifact/10.Docker-Compose/docs.md:
````diff
@@ -0,0 +1,244 @@
+chmod +x $BINDIR/app-gen
+```
+
+### Prepare the deployment
````
```suggestion
### Creating a deployment
```

estape11 commented on 06.Artifact-creation/01.Create-an-Artifact/10.Docker-Compose/docs.md:
````diff
@@ -0,0 +1,244 @@
+
+!!! To quickly verify the required dependencies are installed on your device, run
+!!! the following:
+!!! ```bash
+!!! mender --version
+!!! docker --version
+!!! docker compose version
+!!! xdelta3 -V
+!!! jq --version
+!!! ```
````
```suggestion

```bash
mender --version
docker --version
docker compose version
xdelta3 -V
jq --version
```

```

estape11 replied:
This looks nicer

alfrunes replied:
Removed the colored box, but replaced it with a regular indent (`>`).

estape11 commented:
@alfrunes we added the final minor corrections. We reviewed, re-reviewed and tested and we agreed it is complete.
Thank you for your openness and patience throughout this review.

TheMeaningfulEngineer reviewed (commented):
Whatever @estape11 decides.

alfrunes commented:
I added the changes based on your feedback. I also replaced the paragraph that was moved to the beginning. You can see the diff from the previous version [here](https://github.com/mendersoftware/mender-docs/compare/0843cc79a9748a89f295be1ad60b822b34442eb2..96ea55e6b1897be9dcc94d1863f008ad50e21973).

estape11 reviewed (approved):
We are all set, thanks @alfrunes

alfrunes commented:
> We are all set, thanks @alfrunes

Thanks @estape11 and the CE team :smile_cat: 
@TheMeaningfulEngineer I'm still blocked from merging, I need your approval as well.
